tensorcircuit/templates/lattice.py

```diff
@@ -0,0 +1,886 @@
+# -*- coding: utf-8 -*-
+"""
+The lattice module for defining and manipulating lattice geometries.
+"""
+import abc
+from typing import (
+    Any,
+    Dict,
+    Hashable,
+    Iterator,
+    List,
+    Optional,
+    Tuple,
+    Union,
+    TYPE_CHECKING,
+)
+
+import matplotlib.pyplot as plt
+
+import numpy as np
+
+if TYPE_CHECKING:
+    import matplotlib.pyplot as plt
+    import matplotlib.axes
+    from mpl_toolkits.mplot3d import Axes3D
+
+
+# --- Type Aliases for Readability ---
+SiteIndex = int
+SiteIdentifier = Hashable
+Coordinates = np.ndarray[Any, Any]
+NeighborMap = Dict[SiteIndex, List[SiteIndex]]
+
+
+class AbstractLattice(abc.ABC):
+    """Abstract base class for describing lattice systems.
+
+    This class defines the common interface for all lattice structures,
+    providing access to fundamental properties like site information
+    (count, coordinates, identifiers) and neighbor relationships.
+    Subclasses are responsible for implementing the specific logic for
+    generating the lattice points and calculating neighbor connections.
+
+    Args:
+        dimensionality (int): The spatial dimension of the lattice (e.g., 1, 2, 3).
+    """
+
+    def __init__(self, dimensionality: int):
+        """Initializes the base lattice class."""
+        self._dimensionality = dimensionality
+
+        # --- Internal Data Structures (to be populated by subclasses) ---
+        self._indices: List[SiteIndex] = []
+        self._identifiers: List[SiteIdentifier] = []
+        self._coordinates: List[Coordinates] = []
+        self._ident_to_idx: Dict[SiteIdentifier, SiteIndex] = {}
+        self._neighbor_maps: Dict[int, NeighborMap] = {}
+
+    @property
+    def num_sites(self) -> int:
+        """Returns the total number of sites (N) in the lattice."""
+        return len(self._indices)
+
+    @property
+    def dimensionality(self) -> int:
+        """Returns the spatial dimension of the lattice."""
+        return self._dimensionality
+
+    def __len__(self) -> int:
+        """Returns the total number of sites, enabling `len(lattice)`."""
+        return self.num_sites
+
+    # --- Public API for Accessing Lattice Information ---
+
+    def get_coordinates(self, index: SiteIndex) -> Coordinates:
+        """Gets the spatial coordinates of a site by its integer index.
+
+        Args:
```

Comment:
for the format of the docstring, please refer to other files in this repo, e.g., https://github.com/tensorcircuit/tensorcircuit-ng/blob/master/tensorcircuit/quantum.py#L1327-L1359